crates/pipeline_manager/src/auth.rs

```diff
@@ -0,0 +1,368 @@
+    token_use: String,
+
+    // The username. Note: this may not be unique within a user pool.
+    // The sub claim is the appropriate identifier for a user.
+    username: String,
+}
+
+//
+// Follows the guidelines in the following links, except that JWK refreshes are not
+// yet implemented
+//
+// https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-the-access-token.html
+// JWT claims: https://datatracker.ietf.org/doc/html/rfc7519#section-4
+async fn decode_aws_cognito_token(
+    token: &str,
+    configuration: AuthConfiguration,
+) -> Result<Claim, jsonwebtoken::errors::Error> {
+    let header = decode_header(token).unwrap();
+    match header.alg {
+        // AWS Cognito user pools use RS256
+        Algorithm::RS256 => {
+            let keymap = fetch_jwk_keys(&configuration).await;
+            let jwk = match keymap.get(&header.kid.unwrap()) {
+                Some(jwk) => jwk,
+                None => todo!("Implement key refresh"),
+            };
+            let token_data = decode::<AwsCognitoClaim>(token, jwk, &configuration.validation);
+            if let Ok(t) = &token_data {
+                // TODO: aud and client_id may not be the same when using a resource server
+                if !&configuration
+                    .validation
+                    .aud
+                    .unwrap()
+                    .iter()
+                    .any(|aud| *aud == t.claims.client_id)
+                {
+                    return Err(jsonwebtoken::errors::ErrorKind::InvalidAudience.into());
+                }
+            }
+            match token_data {
+                Ok(data) => {
+                    if data.claims.token_use != "access" {
+                        return Err(jsonwebtoken::errors::ErrorKind::InvalidToken.into());
+                    }
+                    Ok(Claim::AwsCognito(data))
+                }
+                Err(jwt_error) => Err(jwt_error),
+            }
+        }
+        _ => Err(jsonwebtoken::errors::ErrorKind::InvalidAlgorithm.into()),
+    }
+}
+
+async fn fetch_jwk_keys(configuration: &AuthConfiguration) -> HashMap<String, DecodingKey> {
+    match &configuration.provider {
+        Provider::AwsCognito(url) => fetch_jwk_aws_cognito_keys(url).await,
+    }
+}
+
+// We don't want to fetch keys on every authentication attempt, so cache the results.
+// TODO: implement periodic refresh
+cached::cached! {
+    KEYS: SizedCache<String, HashMap<String, DecodingKey>> = SizedCache::with_size(5);
+    async fn fetch_jwk_aws_cognito_keys(url: &String) -> HashMap<String, DecodingKey> = {
+        let res = reqwest::get(url)
+                    .await
+                    .unwrap()
+                    .text()
+                    .await
+                    .unwrap();
+        let keys_as_json: Value = serde_json::from_str(&res).unwrap();
+        keys_as_json
+            .get("keys")
+            .unwrap()
+            .as_array()
+            .unwrap()
```

Comment:
more unwraps below

Reply:
I will leave these unwraps in because this means something unexpectedly changed in the JWK format from AWS.

Reply:
I would definitely avoid panic's that can be triggered by external inputs. Why not just throw an error?

Reply:
Okay, turned those into errors now.